File: 0518/1043_거짓말_백준.py
import sys
sys.stdin = open('1043.txt')

def so(n):
    for j in people[n]:
        if p[j] == 0:
            p[j] = 1
            so(j)

N, M = map(int, input().split())

people = [{i} for i in range(N + 1)]
temp = list(map(int, input().split()))[1:]

party = [list(map(int, input().split())) for _ in range(M)]

for i in party:
    i = i[1:]
    for j in i:
        people[j] |= set(i)

p = [0 for _ in range(N + 1)]

for i in temp:
    so(i)

result = 0
for i in party:
    for j in range(1, len(i)):
        if p[i[j]] == 1:
            break
    else:
        result += 1

print(result)

File: 0518/1043_거짓말_백준.js
// 1. python으로는 반복문으로 풀었으므로 unionFind로 풀어봄

const readline = require('readline')
const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
})

input = []

solution = () => {
    const nm = input[0].split(" ").map((el) => parseInt(el))
    let n = nm[0]
    let m = nm[1]
    let know = input[1].split(" ").map(Number)
    if (know.length !== 1){
        know.splice(0, 1)

    }
    parrent = []
    for (let i = 0; i < n + 1; i++) {
        parrent.push(i)
    }
    for (let i = 2; i < m + 2; i++) {
        let temp = input[i].split(" ").map(Number)
        if (temp[0] === 0) continue
        let root = temp[1]
        for (let ps = 1; ps < temp.length; ps++) {
            if (root === find(temp[ps])) continue
            union(root, temp[ps])
        }
    }
    let iKnow = Array(n + 1).fill(false)
    for (i of know) {
        iKnow[find(i)] = true
    }
    result = 0
    for (let i = 2; i < m + 2; i++) {
        let temp = input[i].split(" ").map(Number)
        if (temp[0] === 0) continue
        let lie = true
        for (let j = 1; j < temp.length; j++) {
            if (iKnow[parrent[find(temp[j])]]) {
                lie = false
                break
            }
        }
        if (lie === true) {
            result += 1
        }
    }
    console.log(result)
}

find = x => {
    if (x === parrent[x]) return x;
    parrent[x] = find(parrent[x])
    return parrent[x]
}

union = (x, y) => {
    let fx = find(x)
    let fy = find(y)
    if (fx !== fy) parrent[fy] = fx
}

rl.on('line', line => input.push(line))
.on('close', () => {
    solution()
    process.exit()
})